### PR TITLE
Addon: add headlamp

### PR DESCRIPTION
**Pull request by yolossn**
Signed-off-by: Santhosh Nagaraj S <santhosh@kinvolk.io>

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -164,6 +163,7 @@ require (
 	github.com/hashicorp/go-safetemp v1.0.0 // indirect
 	github.com/hashicorp/go-version v1.2.1 // indirect
 	github.com/hashicorp/hcl v1.0.0 // indirect
+	github.com/hectane/go-acl v0.0.0-20190604041725-da78bae5fc95 // indirect
```
Comment (illume) [inline go.mod:166]: I wonder if the go module files need to be changed?

File: pkg/addons/config.go [hunk 2]
```diff
@@ -197,4 +197,9 @@ var Addons = []*Addon{
 		set:       SetBool,
 		callbacks: []setFn{EnableOrDisableAddon},
 	},
+	{
```
Comment (illume) [inline pkg/addons/config.go:200]: Seems this list is in alphabetical order.

**Review by illume — COMMENTED** [review]
Awesome.

I didn't try it yet (I will in a while!). Been looking forward to using this :)

- Where is the description of the addon? Probably this needs to be added somewhere.
- Why would this be added as well as the dashboard addon? Probably write that it's extensible dashboard.

File: pkg/addons/config.go (hunk 2 above)
Comment (yolossn) [reply pkg/addons/config.go:200]: I don't think so, `csi-hostpath-driver` is after `volumesnapshots`

**Review by joaquimrocha — CHANGES_REQUESTED** [review]
Pretty cool! What would be the commands for users to be able to expose/access Headlamp in the browser? Have you checked whether we can expose an endpoint, so we can `minikube addons open headlamp`?

I'm not sure this is possible but my idea is that new k8s users could set up Headlamp and access it in the browser without any prior knowledge of how to expose services in k8s.

File: pkg/addons/config.go (hunk 2 above)
Comment (illume) [reply pkg/addons/config.go:200]: Indeed. They seem to be alphabetical at the beginning, but not the last 7.

Comment (yolossn) [reply pkg/addons/config.go:200]: I think the order doesn't matter as the list is sorted when it is displayed.
image

**Comment by yolossn** [comment]
> * Where is the description of the addon? Probably this needs to be added somewhere.
The description should be added in the upstream PR.
 
> * Why would this be added as well as the dashboard addon? Probably write that it's extensible dashboard.
Will include this in the upstream PR description 👍🏽

File: deploy/addons/headlamp/README.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
+## Headlamp Addon
+[Headlamp](https://github.com/kinvolk/headlamp) - Headlamp is an easy-to-use and extensible Kubernetes web UI.
```
Comment (yolossn) [inline deploy/addons/headlamp/README.md:2]: Have added documentation for the addon.

File: deploy/addons/headlamp/README.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+### Enable Headlamp on minikube
+
+To enable this addon, simply run:
+```shell script
+minikube addons enable headlamp
+```
+
+Once the addon is enable you can access the Headlamp web UI use the following command.
```
Comment (joaquimrocha) [inline deploy/addons/headlamp/README.md:11]:
```suggestion
Once the addon is enabled, you can access Headlamp's web UI using the following command:
```

File: deploy/addons/headlamp/README.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+```shell script
+minikube service headlamp -n headlamp
+```
+
+To authenticate in Headlamp, fetch the Authentication Token use the following command:
```
Comment (joaquimrocha) [inline deploy/addons/headlamp/README.md:16]:
```suggestion
To authenticate in Headlamp, fetch the Authentication Token using the following command:
```

File: cmd/minikube/cmd/config/enable.go [hunk 6]
```diff
@@ -67,6 +67,29 @@ var addonsEnableCmd = &cobra.Command{
...
 
+`, out.V{"profileArg": tipProfileArg})
+
+		}
+		if addon == "headlamp" {
+			out.Styled(style.Tip, `To access the Headlamp web UI use the following command
+minikube service headlamp -n headlamp
+
+`)
+			out.Styled(style.Tip, `To authenticate in Headlamp, fetch the Authentication Token use the following command:
```
Comment (joaquimrocha) [inline cmd/minikube/cmd/config/enable.go:78]:
```suggestion
			out.Styled(style.Tip, `To authenticate in Headlamp, fetch the Authentication Token using the following command:
```

File: deploy/addons/headlamp/README.md (hunk 3 above)
Comment (joaquimrocha) [inline deploy/addons/headlamp/README.md:2]:
```suggestion
[Headlamp](https://kinvolk.github.io/headlamp) - Headlamp is an easy-to-use and extensible Kubernetes web UI.
```

File: deploy/addons/headlamp/README.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+Headlamp can display more detailed information when metrics-server is installed. To install it, run:
+
+```shell script
+minikube addons enable metrics-server	
+```		
+
+### Testing installation
+
+```shell script
+kubectl get po -n headlamp
```
Comment (joaquimrocha) [inline deploy/addons/headlamp/README.md:32]: Should it be `get pods` (to be a bit more formal).

File: deploy/addons/headlamp/README.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+```
+
+If everything went well you shouldn't get any errors about headlamp being installed in your cluster.
```
Comment (joaquimrocha) [inline deploy/addons/headlamp/README.md:35]:
```suggestion
If everything went well, there should be no errors about Headlamp's installation in your Minikube cluster.
```

File: deploy/addons/headlamp/README.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+### Disable headlamp
+To disable this addon, simply run:
```
Comment (joaquimrocha) [inline deploy/addons/headlamp/README.md:38]: Add a newline under each of these lines.

File: deploy/addons/headlamp/README.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
+## Headlamp Addon
```
Comment (joaquimrocha) [inline deploy/addons/headlamp/README.md:1]: Add a newline under this.

File: deploy/addons/headlamp/README.md (hunk 3 above)
Comment (joaquimrocha) [reply deploy/addons/headlamp/README.md:2]: BTW, why not just add the link + the description: [Headlamp](...) is an easy-to-use ....

Comment (joaquimrocha) [reply deploy/addons/headlamp/README.md:2]: I have also noticed that apparently other addons don't have a README.md file, so not sure if they will accept it for us.

File: site/content/en/docs/handbook/addons/headlamp.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
+---
+title: "Using Headlamp Addon"
+linkTitle: "Headlamp"
+weight: 1
+date: 2022-06-08
+---
+
+
+## Headlamp Addon
+[Headlamp](https://github.com/kinvolk/headlamp) - Headlamp is an easy-to-use and extensible Kubernetes web UI.
```
Comment (joaquimrocha) [inline site/content/en/docs/handbook/addons/headlamp.md:10]: check the comments on the README, which should apply here too.

**Review by joaquimrocha — CHANGES_REQUESTED** [review]
It's looking very good. I found some issues related to language, so left some comments.
Also, I see you added the volume claim but I haven't seen instructions on how to set up the plugins. If the instructions should be in headlamp's docs, maybe we shouldn't have the volume claim here and just have instructions on how to set up plugins in a generic (in-cluster) way and point minikube's users there on a successful addon enabling.

**Comment by illume** [comment]
> > The [dashboard addon documentation](https://github.com/kubernetes/minikube/blob/master/site/content/en/docs/handbook/dashboard.md#getting-just-the-dashboard-url) had something interesting... it had a --url option. Perhaps we could also have that to match the dashboard addon?

> This looks custom made for the dashboard command, I don't think something like this is possible unless we propose a headlamp command.

Seems to be some other optional addons have a command. Like this gvisor one: https://github.com/kubernetes/minikube/pull/3399 In perhaps a future PR I think it's worth asking about adding an optional headlamp cmd... perhaps on a future PR. Being able to run code could be useful for making some of these things run as smoothly as `minkube dashboard`. 

ps. The I noticed the service command works with --url similar to the `minikube dashboard --url`.

File: site/content/en/docs/handbook/addons/headlamp.md (hunk 11 above)
Comment (yolossn) [reply site/content/en/docs/handbook/addons/headlamp.md]: Removed the README and added the link to the docs in the addon description.

File: cmd/minikube/cmd/config/enable.go [hunk 12]
```diff
@@ -67,6 +67,29 @@ var addonsEnableCmd = &cobra.Command{
...
+
+export SECRET=$(kubectl get secrets --namespace headlamp -o custom-columns=":metadata.name" | grep "headlamp-token")
+kubectl get secret $SECRET --namespace headlamp --template=\{\{.data.token\}\} | base64 --decode
```
[the remark this answers is not in the document]
Comment (yolossn) [reply cmd/minikube/cmd/config/enable.go:81]: I think it makes sense to update the documentation as it involves the user to copy the headlamp-token name, whereas this solution can be run directly.

**Review by illume — APPROVED** [review]
lgtm 🎈🎉

**Pull request closed without merging**